From 419361fcae0929dd1e99e7ba1890a982de318039 Mon Sep 17 00:00:00 2001
From: Hans de Goede <johannes.goede@oss.qualcomm.com>
Date: Fri, 13 Mar 2026 19:13:57 +0100
Subject: [PATCH 13/16] feat(dracut): add module to load Qualcomm ADSP module
 pre-udev

Cherry picked from upstream commit 647a685d3567 ("feat(dracut): add module
to load Qualcomm ADSP module pre-udev"). Upstream PR (merged):
https://github.com/dracut-ng/dracut-ng/pull/2302

On Qualcomm sc8280xp and x1e laptops the kernel reboots the ADSP with new
firmware because the BIOS loads ADSP firmware with limited functionality
without sound or battery charge/status reading support.

Unfortunately the ADSP also controls the TCPM (Type-C Port Manager) and
rebooting the ADSP also resets the TCPM, causing any USB devices connected
over Type-C ports to get disconnected as all devices on the USB bus are
removed and re-enumerated.

This breaks booting from USB-drives as the drive gets disconnected and
re-enumerated as a new block device, leaving any filesystems mounted
before the ADSP reset without any backing device.

To workaround this add a module to load the ADSP driver from a pre-udev
hook so that the USB re-enumeration happens before the rootfs is mounted.

Signed-off-by: Hans de Goede <johannes.goede@oss.qualcomm.com>
---
Changes in v5:
- Sort `inst_multiple` arguments alphabetically

Changes in v4:
- Change host-only check to [[ $hostonly_mode == "strict" ]] so that
  30-qcom-adsp-pre-udev.sh gets installed in sloppy hostonly mode
- Add modprobe to list of binaries to install from install()

Changes in v3:
- Add "inst grep" to install()

Changes in v2:
- Directly use $DRACUT_ARCH instead of defining _arch local var
- Add the new module to labeler.yml and core.adoc
---
 doc_site/modules/ROOT/pages/modules/core.adoc |  4 ++
 modules.d/70qcom-adsp/module-setup.sh         | 40 +++++++++++++++++++
 modules.d/70qcom-adsp/qcom-adsp-pre-udev.sh   |  9 +++++
 3 files changed, 53 insertions(+)
 create mode 100755 modules.d/70qcom-adsp/module-setup.sh
 create mode 100755 modules.d/70qcom-adsp/qcom-adsp-pre-udev.sh

diff --git a/doc_site/modules/ROOT/pages/modules/core.adoc b/doc_site/modules/ROOT/pages/modules/core.adoc
index 42538dea..9f2dffc7 100644
--- a/doc_site/modules/ROOT/pages/modules/core.adoc
+++ b/doc_site/modules/ROOT/pages/modules/core.adoc
@@ -201,6 +201,10 @@ code, there are no specific types or categories for dracut modules.
 | polls CD-ROM
 |
 
+| qcom-adsp
+| Load Qualcomm ADSP module pre-udev
+|
+
 | rescue
 | utilities for rescue mode (such as ping, ssh, vi, fsck.*)
 | utils
diff --git a/modules.d/70qcom-adsp/module-setup.sh b/modules.d/70qcom-adsp/module-setup.sh
new file mode 100755
index 00000000..a3d05057
--- /dev/null
+++ b/modules.d/70qcom-adsp/module-setup.sh
@@ -0,0 +1,40 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: nil; sh-basic-offset: 4; -*-
+# ex: ts=8 sw=4 sts=4 et filetype=sh
+
+# On Qualcomm sc8280xp and x1e laptops the kernel reboots the ADSP with new
+# firmware because the BIOS loads ADSP firmware with limited functionality
+# without sound or battery charge/status reading support.
+#
+# Unfortunately the ADSP also controls the TCPM (Type-C Port Manager) and
+# rebooting the ADSP also resets the TCPM, causing any USB devices connected
+# over Type-C ports to get disconnected as all devices on the USB bus are
+# removed and re-enumerated.
+#
+# This breaks booting from USB-drives as the drive gets disconnected and
+# re-enumerated as a new block device, leaving any filesystems mounted
+# before the ADSP reset without any backing device.
+#
+# To workaround this, this module load the ADSP driver from a pre-udev hook
+# so that the USB re-enumeration happens before the rootfs is mounted.
+
+# called by dracut
+check() {
+    [ "$DRACUT_ARCH" = "aarch64" ] || return 1
+    if [[ $hostonly_mode == "strict" ]]; then
+        grep -q -E 'qcom,sc8280xp-adsp-pas|qcom,x1e80100-adsp-pas' \
+            /sys/bus/platform/devices/*.remoteproc/modalias 2> /dev/null || return 1
+    fi
+    return 0
+}
+
+# called by dracut
+installkernel() {
+    instmods qcom_q6v5_pas
+}
+
+# called by dracut
+install() {
+    inst_multiple grep modprobe
+    inst_hook pre-udev 30 "$moddir/qcom-adsp-pre-udev.sh"
+}
diff --git a/modules.d/70qcom-adsp/qcom-adsp-pre-udev.sh b/modules.d/70qcom-adsp/qcom-adsp-pre-udev.sh
new file mode 100755
index 00000000..2c408428
--- /dev/null
+++ b/modules.d/70qcom-adsp/qcom-adsp-pre-udev.sh
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+# Ensure ADSP module is loaded and Type-C USB busses are reset before
+# filesystems are mounted
+
+if grep -q -E 'qcom,sc8280xp-adsp-pas|qcom,x1e80100-adsp-pas' \
+    /sys/bus/platform/devices/*.remoteproc/modalias 2> /dev/null; then
+    modprobe qcom_q6v5_pas
+fi
-- 
2.55.0

